### DISCUSSION TITLE
Dynamic colors not compiling properly

### DISCUSSION THREAD

**Discussion by BigRaj** [tailwindlabs/tailwindcss#10315] [Help] [answered]
I have a react component that passes in a static color name, for instance "gray"  `<Button background="gray" />`

The component itself applies the color as `className='bg-${color}-500 hover:bg-${color}-300'`.  The hover element does not apply.  

When applying classes without variables passed in, the hover function works correctly.  Is there something that can be done to the config to account for this?

**Comment by wongjn** [comment, marked as answer]
You could look at [`safelist`](https://tailwindcss.com/docs/content-configuration#safelisting-classes) and add the possible classes used there.

However, the root concept of why this does not work is because [you cannot dynamically build Tailwind class names](https://tailwindcss.com/docs/content-configuration#dynamic-class-names). Tailwind scans source files as plain text strings and does not run code, hence it does not "see" `bg-gray-500` since it never encounters that string as-is (think of it like the file is a `.txt` file instead of a `.jsx` file).

**Reply by BigRaj** [reply]
I figured it had something with to do with how tailwind ingests the class names on compilation.  safelist looks like it does what I'm trying to do for one site, but the component color prefixes may change based on customer theme names.  

Honestly it sounds like a business decision type of thing to refactor the component to work using arbitrary names like "primary" and "secondary" instead of actual color names.

**Comment by adamwathan** [comment]
Generally the solution here is to map incoming props to a list of complete class names rather than using interpolation:

```jsx
function Button({ background, children }) {
  const backgroundVariants = {
    gray: 'bg-gray-500 hover:bg-gray-300',
    blue: 'bg-blue-600 hover:bg-blue-500',
    // ...
  }

  return (
    <button className={`${backgroundVariants[background]} ...`}>
      {children}
    </button>
  )
}
```

This has the added benefit of letting you use different shades for each variant as well, or making other changes like sometimes using white text and other times using dark text depending on the button color.
